File: supabase/functions/_shared/embedding.ts
/** Parse embedding from DB (string or array) to number[] */
export function parseEmbedding(val: unknown): number[] {
  if (Array.isArray(val)) return val.map(Number)
  if (typeof val === 'string') {
    const s = val.replace(/^\[|\]$/g, '').trim()
    return s ? s.split(',').map(Number) : []
  }
  return []
}

/** Cosine similarity between two 128-d vectors */
export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== 128 || b.length !== 128) return 0
  let dot = 0
  let na = 0
  let nb = 0
  for (let i = 0; i < 128; i++) {
    dot += a[i] * b[i]
    na += a[i] * a[i]
    nb += b[i] * b[i]
  }
  return na && nb ? dot / (Math.sqrt(na) * Math.sqrt(nb)) : 0
}

File: backend/app/services/embedding.py
"""Face embedding utilities - cosine similarity, parsing."""
import numpy as np


def parse_embedding(val: list | str) -> list[float]:
    """Parse embedding from DB (string or array) to list of floats."""
    if isinstance(val, list):
        return [float(x) for x in val]
    if isinstance(val, str):
        s = val.strip("[]").strip()
        return [float(x) for x in s.split(",")] if s else []
    return []


def cosine_similarity(a: list[float], b: list[float]) -> float:
    """Cosine similarity between two 128-d vectors."""
    if len(a) != 128 or len(b) != 128:
        return 0.0
    arr_a = np.array(a, dtype=np.float64)
    arr_b = np.array(b, dtype=np.float64)
    dot = np.dot(arr_a, arr_b)
    na = np.linalg.norm(arr_a)
    nb = np.linalg.norm(arr_b)
    if na == 0 or nb == 0:
        return 0.0
    return float(dot / (na * nb))
